### PR TITLE
feat: Implement V2 FBCaching and Optimize Existing FBCache

### PR DESCRIPTION
**Pull request by Bluear7878**
## Description
This PR is a new V2 version of FB caching, complete with test cases and examples. Additionally, it includes optimizations for the existing FBCache implementation.

Key Changes
Implemented V2 FB Caching:

Introduced a new FBCacheV2 with a revised caching logic.

Added test cases.

Included an example script of FBCacheV2.

Optimized Existing FBCache:

#### Without Cache (Baseline)
- **V1 Baseline**: ~27.6s (50 steps)
- **V2 Baseline**: ~29.0s (50 steps)

#### With Cache (multi_threshold=0.12)
```
V1 FBCache Performance:
├── Run 1: 9.70s
├── Run 2: 9.83s
├── Run 3: 9.92s
└── Average: 9.82s (196.3ms/step)

V2 FBCache Performance:
├── Run 1: 10.74s
├── Run 2: 10.81s
├── Run 3: 10.85s
└── Average: 10.80s (216.0ms/step)
```

DoubleFBCache Performance comparsion

### Key Findings

| Single_Threshold | V1 Time (s) | V2 Time (s) | Difference|
|-----------|:-----------:|:-----------:|:----------:|
| **0.08** | 22.02 | 23.37 | +1.35s (+6.1%) |
| **0.10** | 18.18 | 18.80 | +0.61s (+3.4%) |
| **0.12** | 13.98 | 14.53 | +0.55s (+4.0%) |
| **0.15** | 13.27 | 13.96 | +0.69s (+5.2%) |
| **0.20** | 13.28 | 13.95 | +0.68s (+5.1%) |

### Cache Effectiveness

| Method| Without Cache | With Cache | Speedup | Effectiveness |
|----------------|:-------------:|:----------:|:-------:|:-------------:|
| **V1** | 27.6s | 9.82s | **17.78s** | **64.4%** |
| **V2** | 29.0s | 10.80s | **18.20s** | **62.8%** |

FBCache refactoring and optimzation
**FBCache.py (Update) delivers marginal but consistent performance improvements** over the original FBCache.py by eliminating GPU-CPU synchronization overhead. The optimized version is **1.2% faster** in realistic workflows while maintaining identical functionality.

| Metric | Original | Optimized | Improvement |
|--------|:--------:|:---------:|:-----------:|
| **Similarity Check (avg)** | 3.98ms | 3.98ms | ~0% |
| **Cache Operations** | 0.0023ms | 0.0022ms | 4.3% |
| **Workflow Simulation** | 10.34ms | 10.22ms | **1.2%**  |
| **GPU-CPU Sync Overhead** | 0.10ms | 0ms | **100%**  |

### CODE REVIEW

**Comment by lmxyy** [comment]
Could you provide the reference cache implementation in diffusers?

**Comment by Bluear7878** [comment]
> Could you provide the reference cache implementation in diffusers?

I looked into implementing caching in diffusers.
FasterCache: [link](https://huggingface.co/docs/diffusers/main/en/optimization/cache)

FBCache (in ParaAttn): [link](https://huggingface.co/docs/diffusers/main/en/optimization/para_attn)

DeepCache: [link](https://huggingface.co/docs/diffusers/main/en/optimization/deepcache)

Should I align my current implementation with the style and structure of these existing caches?
Any specific one I should prioritize as a reference?

**Comment by lmxyy** [comment]
> > Could you provide the reference cache implementation in diffusers?
> 
> I looked into implementing caching in diffusers. FasterCache: [link](https://huggingface.co/docs/diffusers/main/en/optimization/cache)
> 
> FBCache (in ParaAttn): [link](https://huggingface.co/docs/diffusers/main/en/optimization/para_attn)
> 
> DeepCache: [link](https://huggingface.co/docs/diffusers/main/en/optimization/deepcache)
> 
> Should I align my current implementation with the style and structure of these existing caches? Any specific one I should prioritize as a reference?

Yeah, I think maybe we need to align the current implementation if we would like to scale up the codebase and support more cache mechanisms in the future. Better check how diffusers maintain the codebase.

**Comment by Bluear7878** [comment]
> > > Could you provide the reference cache implementation in diffusers?
> > 
> > 
> > I looked into implementing caching in diffusers. FasterCache: [link](https://huggingface.co/docs/diffusers/main/en/optimization/cache)
> > FBCache (in ParaAttn): [link](https://huggingface.co/docs/diffusers/main/en/optimization/para_attn)
> > DeepCache: [link](https://huggingface.co/docs/diffusers/main/en/optimization/deepcache)
> > Should I align my current implementation with the style and structure of these existing caches? Any specific one I should prioritize as a reference?
> 
> Yeah, I think maybe we need to align the current implementation if we would like to scale up the codebase and support more cache mechanisms in the future. Better check how diffusers maintain the codebase.

I got it. i will make those caches to the format that diffuser supports

File: nunchaku/caching/utils_v2.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+.. note::
+   V2 implementation with standalone functions for improved modularity.
+"""
+
+from typing import Any, Dict, Optional, Union
+
+import torch
+from diffusers.models.modeling_outputs import Transformer2DModelOutput
+
+from nunchaku.caching.FBCache import check_and_apply_cache
```
Comment (lmxyy) [inline nunchaku/caching/utils_v2.py:30]: For Python file, we'd better call it `fbcache` instead of `FBCache`.

File: nunchaku/caching/FBCache.py
Comment (lmxyy) [inline nunchaku/caching/FBCache.py:1]: rename the file to `fbcache.py`.

File: nunchaku/caching/diffusers_adapters/flux_v2.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+def apply_cache_on_transformer(
+    transformer: NunchakuFluxTransformer2DModelV2,
+    *,
+    use_double_fb_cache: bool = False,
+    residual_diff_threshold: float = 0.12,
+    residual_diff_threshold_multi: float | None = None,
+    residual_diff_threshold_single: float | None = None,
+):
+    """
+    Apply caching to transformer by replacing its forward method.
```
Comment (lmxyy) [inline nunchaku/caching/diffusers_adapters/flux_v2.py:24]: you need to specify what these arguments mean.

**Review by lmxyy — APPROVED** [review]
